File: spoj/SVADA-binary-search-monkey/sol.c
// Binary search working solution in C

#include <stdio.h>

int countCocos(int N, int T, int B[], int C[]) {
	int c = 0;
	for(int i=0; i<N; i++) {
		if(T >= B[i]){
			c += 1 + ((T - B[i]) / C[i]);			
		}
	}
	return c;

}

int main() {

	// Input
	int T, N, M;
	int A[100], B[100], C[100], D[100];
	scanf("%d", &T);
	scanf("%d", &N);
	for(int i=0; i<N; i++){
		scanf("%d %d", &A[i], &B[i]);
	}
	scanf("%d", &M);
	for(int i=0; i<M; i++){
		scanf("%d %d", &C[i], &D[i]);
	}

	int lo = 0;
	int hi = T;
	int res = -1;
	while(hi >= lo) {
		int mid = (hi + lo) / 2;
		int cocosA = countCocos(N, mid, A, B);
		int cocosB = countCocos(M, T - mid, C, D);
		// printf("debug %d %d %d %d %d\n", lo, mid, hi, cocosA, cocosB);
		if(cocosA > cocosB) {
			hi = mid - 1;
		} else {
			res = mid;
			lo = mid + 1;
		}
	}
	printf("%d\n", res);
	return 0;
}

File: spoj/SVADA-binary-search-monkey/sol.py
# Binary search working solution in Python 3

from functools import reduce

class Data:

	def __init__(self):
		self.tot = int(input())
		self.numA = int(input())
		self.monksA = []
		for _ in range(self.numA):
			self.monksA.append(Monkey(*[int(x) for x in input().split()]))
		self.numB = int(input())
		self.monksB = []
		for _ in range(self.numB):
			self.monksB.append(Monkey(*[int(x) for x in input().split()]))

class Monkey:

	def __init__(self, a, b):
		self.a = a
		self.b = b

	def get_cocos(self, t):
		cocos = 1 if t >= self.a else 0
		cocos += int(max(0, t - self.a) / self.b)
		return cocos

def get_all_cocos(monks, t):
	return reduce(lambda a,b: a+b, map(lambda m: m.get_cocos(t), monks))

def solve(d):
	lo = 0
	hi = d.tot
	res = None
	while hi >= lo:
		mid = int((hi + lo) / 2)
		cocosA = get_all_cocos(d.monksA, mid)
		cocosB = get_all_cocos(d.monksB, d.tot-mid)
		# print(lo, mid, hi, cocosA, cocosB)
		if cocosA > cocosB:
			hi = mid - 1
		else:
			res = mid
			lo = mid + 1
	if res is None:
		raise Exception('Should not have happened')
	else:
		print(res)

d = Data()
solve(d)